code/datums/dna.dm

```diff
@@ -120,6 +123,48 @@
 			. += random_string(DNA_BLOCK_SIZE,GLOB.hex_characters)
 	return .
 
+/datum/dna/proc/generate_uni_features()
+	. = ""
+	var/list/L = new /list(DNA_FEATURE_BLOCKS)
+
+	if(features["mcolor"])
+		L[DNA_MCOLOR_BLOCK] = sanitize_hexcolor(features["mcolor"])
+	if(features["ethcolor"])
+		L[DNA_ETHCOLOR_BLOCK] = sanitize_hexcolor(features["ethcolor"])
+	if(features["body_markings"])
+		L[DNA_LIZARD_MARKINGS_BLOCK] = construct_block(GLOB.body_markings_list.Find(features["body_markings"]), GLOB.body_markings_list.len)
+	if(features["tail_lizard"])
+		L[DNA_LIZARD_TAIL_BLOCK] = construct_block(GLOB.tails_list_lizard.Find(features["tail_lizard"]), GLOB.tails_list_lizard.len)
+	if(features["snout"])
+		L[DNA_SNOUT_BLOCK] = construct_block(GLOB.snouts_list.Find(features["snout"]), GLOB.snouts_list.len)
+	if(features["horns"])
+		L[DNA_HORNS_BLOCK] = construct_block(GLOB.horns_list.Find(features["horns"]), GLOB.horns_list.len)
+	if(features["frills"])
+		L[DNA_FRILLS_BLOCK] = construct_block(GLOB.frills_list.Find(features["frills"]), GLOB.frills_list.len)
+	if(features["spines"])
+		L[DNA_SPINES_BLOCK] = construct_block(GLOB.spines_list.Find(features["spines"]), GLOB.spines_list.len)
+	if(features["tail_human"])
+		L[DNA_HUMAN_TAIL_BLOCK] = construct_block(GLOB.tails_list_human.Find(features["tail_human"]), GLOB.tails_list_human.len)
+	if(features["ears"])
+		L[DNA_EARS_BLOCK] = construct_block(GLOB.ears_list.Find(features["ears"]), GLOB.ears_list.len)
+	if(features["moth_wings"] != "Burnt Off")
+		L[DNA_MOTH_WINGS_BLOCK] = construct_block(GLOB.moth_wings_list.Find(features["moth_wings"]), GLOB.moth_wings_list.len)
+	if(features["moth_antennae"] != "Burnt Off")
+		L[DNA_MOTH_ANTENNAE_BLOCK] = construct_block(GLOB.moth_antennae_list.Find(features["moth_antennae"]), GLOB.moth_antennae_list.len)
+	if(features["moth_markings"])
+		L[DNA_MOTH_MARKINGS_BLOCK] = construct_block(GLOB.moth_markings_list.Find(features["moth_markings"]), GLOB.moth_markings_list.len)
+	if(features["caps"])
+		L[DNA_CAPS_BLOCK] = construct_block(GLOB.caps_list.Find(features["caps"]), GLOB.caps_list.len)
+	if(features["tail_monkey"])
+		L[DNA_MONKEY_TAIL_BLOCK] = construct_block(GLOB.tails_list_monkey.Find(features["tail_monkey"]), GLOB.tails_list_monkey.len)
+
+	for(var/i=1, i<=DNA_FEATURE_BLOCKS, i++)
+		if(L[i])
+			. += L[i]
+		else
+			. += random_string(DNA_BLOCK_SIZE,GLOB.hex_characters)
+	return .
```

Comment:
```suggestion
/datum/dna/proc/generate_uni_features()
	var/list/data = list()
	
	for(var/feature in list())
	
	var/list/L = new /list(DNA_FEATURE_BLOCKS)

	if(features["mcolor"])
		L[DNA_MCOLOR_BLOCK] = sanitize_hexcolor(features["mcolor"])
	if(features["ethcolor"])
		L[DNA_ETHCOLOR_BLOCK] = sanitize_hexcolor(features["ethcolor"])
	if(features["body_markings"])
		L[DNA_LIZARD_MARKINGS_BLOCK] = construct_block(GLOB.body_markings_list.Find(features["body_markings"]), GLOB.body_markings_list.len)
	if(features["tail_lizard"])
		L[DNA_LIZARD_TAIL_BLOCK] = construct_block(GLOB.tails_list_lizard.Find(features["tail_lizard"]), GLOB.tails_list_lizard.len)
	if(features["snout"])
		L[DNA_SNOUT_BLOCK] = construct_block(GLOB.snouts_list.Find(features["snout"]), GLOB.snouts_list.len)
	if(features["horns"])
		L[DNA_HORNS_BLOCK] = construct_block(GLOB.horns_list.Find(features["horns"]), GLOB.horns_list.len)
	if(features["frills"])
		L[DNA_FRILLS_BLOCK] = construct_block(GLOB.frills_list.Find(features["frills"]), GLOB.frills_list.len)
	if(features["spines"])
		L[DNA_SPINES_BLOCK] = construct_block(GLOB.spines_list.Find(features["spines"]), GLOB.spines_list.len)
	if(features["tail_human"])
		L[DNA_HUMAN_TAIL_BLOCK] = construct_block(GLOB.tails_list_human.Find(features["tail_human"]), GLOB.tails_list_human.len)
	if(features["ears"])
		L[DNA_EARS_BLOCK] = construct_block(GLOB.ears_list.Find(features["ears"]), GLOB.ears_list.len)
	if(features["moth_wings"] != "Burnt Off")
		L[DNA_MOTH_WINGS_BLOCK] = construct_block(GLOB.moth_wings_list.Find(features["moth_wings"]), GLOB.moth_wings_list.len)
	if(features["moth_antennae"] != "Burnt Off")
		L[DNA_MOTH_ANTENNAE_BLOCK] = construct_block(GLOB.moth_antennae_list.Find(features["moth_antennae"]), GLOB.moth_antennae_list.len)
	if(features["moth_markings"])
		L[DNA_MOTH_MARKINGS_BLOCK] = construct_block(GLOB.moth_markings_list.Find(features["moth_markings"]), GLOB.moth_markings_list.len)
	if(features["caps"])
		L[DNA_CAPS_BLOCK] = construct_block(GLOB.caps_list.Find(features["caps"]), GLOB.caps_list.len)
	if(features["tail_monkey"])
		L[DNA_MONKEY_TAIL_BLOCK] = construct_block(GLOB.tails_list_monkey.Find(features["tail_monkey"]), GLOB.tails_list_monkey.len)

	for(var/i in 1 to DNA_FEATURE_BLOCKS)
		data += (L[i] || random_string(DNA_BLOCK_SIZE,GLOB.hex_characters))
	
	return data.Join()
```

* Use lists and the `Join()` method to collect text data. BYOND text addition is rather poorly implemented.
* Prefer to use list-loops instead of C-loops if you don't care about modifying the index. They're both more common and somewhat cleaner, and don't check whether the index has changed every single loop (and instead work with a copy of the initial range/list).
* No need to `return .`, as `.` is precisely defined as the value that will be returned once the proc ends. Regardless, had to make a local variable to type cast it.